lib/dpl/provider/pages.rb

```diff
@@ -44,34 +53,140 @@ def slug
         options.fetch(:repo) { context.env['TRAVIS_REPO_SLUG'] }
       end
 
+      def keep_history
+        options.fetch(:keep_history) { false }
+      end
+
+      def allow_empty_commit
+        options.fetch(:allow_empty_commit) { false }
+      end
+
+      def verbose
+        # Achtung! Never verbosify git, since it may expose user's token.
+        options.fetch(:verbose) { false }
+      end
+
       def check_auth
       end
 
       def needs_key?
         false
       end
 
-      def github_deploy
-        context.shell 'rm -rf .git > /dev/null 2>&1'
-        context.shell "touch \"deployed at `date` by #{@gh_name}\""
-        context.shell 'git init' or raise 'Could not create new git repo'
+      def print_step(msg)
+        $stderr.puts msg if @verbose
+      end
+
+      def github_pull(target_dir)
+        print_step "Trying to clone a single branch #{@target_branch} from existing repo..."
+        unless context.shell "git clone --quiet --branch='#{@target_branch}' --single-branch '#{@gh_remote_url}' '#{target_dir}' &>/dev/null"
```

Comment:
Consider [`--depth=1`](https://git-scm.com/docs/git-clone#git-clone---depthltdepthgt) to save time when the gh-pages branch gets long. That'll also imply `--single-branch`.

Reply:
@jyasskin I new about this option and I don't remember why I decided not to use it back then. I'll replace it as suggested.